android/engine/src/main/java/org/smartregister/fhircore/engine/task/FhirCarePlanGenerator.kt

```diff
@@ -219,4 +208,98 @@ constructor(
+      require(it.kind == PlanDefinition.ActionConditionKind.APPLICABILITY) {
+        "PlanDefinition.action.kind=${it.kind} not supported"
+      }
+
+      require(it.expression.language == Expression.ExpressionLanguage.TEXT_FHIRPATH.toCode()) {
+        "PlanDefinition.expression.language=${it.expression.language} not supported"
+      }
+
+      fhirPathEngine.evaluateToBoolean(focus, root, base, it.expression.expression)
+    }
+
+  private fun PlanDefinition.PlanDefinitionActionComponent.activityDefinition(
+    planDefinition: PlanDefinition
+  ) =
+    planDefinition.contained.filter { it.resourceType == ResourceType.ActivityDefinition }.first {
+      it.logicalId == this.definitionCanonicalType.value
+    } as
+      ActivityDefinition
+
+  private fun PlanDefinition.PlanDefinitionActionComponent.taskPeriods(
+    definition: ActivityDefinition,
+    carePlan: CarePlan
+  ): List<Period> {
+    return when {
+      definition.hasDosage() -> extractTaskPeriodsFromDosage(definition.dosage, carePlan)
+      definition.hasTiming() && !definition.hasTimingTiming() ->
+        throw IllegalArgumentException(
+          "Timing component should only be Timing. Can not handle ${timing.fhirType()}"
+        )
+      else -> extractTaskPeriodsFromTiming(definition.timingTiming, carePlan)
+    }
+  }
+
+  private fun extractTaskPeriodsFromTiming(timing: Timing, carePlan: CarePlan): List<Period> {
+    val taskPeriods = mutableListOf<Period>()
+    // TODO handle properties used by older plandefintions. If any plan definition is using
+    //  countMax, frequency, durationUnit of hour, consider as non compliant and assume
+    //  all handling would be done into structure map. Once all plan definitions are using
+    //  recommended approach and using plan definitions properly change line below to use
+    //  timing.repeat.count only
+    val count =
+      if (timing.repeat.hasFrequency() ||
+          timing.repeat.hasCountMax() ||
+          timing.repeat.durationUnit?.equals(UnitsOfTime.H) == true
+      )
+        0
+      else timing.repeat.count
+    val periodExpression = timing.extractFhirpathPeriod()
+    val durationExpression = timing.extractFhirpathDuration()
+
+    // offset date for current task period; careplan start if all tasks generated at once
+    // otherwise today means that tasks are generated on demand
+    var offsetDate: BaseDateTimeType = DateType(if (count > 0) carePlan.period.start else Date())
+
+    var i = 1 // task index
+    do {
+      if (periodExpression.isNotBlank())
+        evaluateToDate(offsetDate, "\$this + $periodExpression")?.let { offsetDate = it }
+
+      Period()
+        .apply {
+          start = offsetDate.value
+
+          end =
+            if (durationExpression.isNotBlank())
+              evaluateToDate(offsetDate, "\$this + $durationExpression")?.value
+            else carePlan.period.end
+        }
+        .also { taskPeriods.add(it) }
+
+      i++
+    } while (count >= i)
```

Comment:
can go away if we use a for loop

Reply:
An activity can have a task without repeat schedule (Timing component). The reason is that if there is an activity it has to generate atleast one task (irrespective of whether timing component is defined or resolves to repeat schedule). So its a must do {} and repeat while(condition)